### ISSUE TITLE
v3.0.1 raises an error

### ISSUE THREAD

**Issue by mikearnaldi** [LeDDGroup/typescript-transform-paths#118]
After #114 fixes the following presents:

```
yarn test packages/system/test/
yarn run v1.22.10
$ jest --detectOpenHandles packages/system/test/
Failed to resolve output name for /home/ma/os/slicing/core/packages/system/src/Utils/union.ts. Please report a GH issue at: https://github.com/LeDDGroup/typescript-transform-paths/issues
Failed to resolve output name for /home/ma/os/slicing/core/packages/system/src/Collections/Immutable/Chunk/core.ts. Please report a GH issue at: https://github.com/LeDDGroup/typescript-transform-paths/issues
Failed to resolve output name for /home/ma/os/slicing/core/packages/system/src/Collections/Immutable/Chunk/definition.ts. Please report a GH issue at: https://github.com/LeDDGroup/typescript-transform-paths/issues
Failed to resolve output name for /home/ma/os/slicing/core/packages/system/src/Collections/Immutable/Chunk/api/filterMap.ts. Please report a GH issue at: https://github.com/LeDDGroup/typescript-transform-paths/issues
Failed to resolve output name for /home/ma/os/slicing/core/packages/system/src/Collections/Immutable/Chunk/api/compact.ts. Please report a GH issue at: https://github.com/LeDDGroup/typescript-transform-paths/issues
```

same repo as the previous issue for repro

**Comment by nonara** [comment]
Is the output correct?

**Comment by nonara** [comment]
Looking into this. This plugin is primarily used for rewriting actual tsc output files. It looks like you're using it within jest, which uses a hosted version of typescript and applies some other magic, so results can't be guaranteed.

In this case, it looks like the actual typescript compiler API function `getOutputFileNames` is failing to resolve them because the files in question aren't directly referenced from the jest project file. 

As far as that goes, no need to change anything. I'm going to look into this further and see if it's something that can be corrected on my end.

However, have you looked at the [pathsToModuleNameMapper](https://kulshekhar.github.io/ts-jest/docs/getting-started/paths-mapping/#jest-config-with-helper) utility within `ts-jest`? If you can, that is the route I would advise going for jest mapping. It is likely to be faster, also.

**Comment by nonara** [comment]
I've implemented an alternate pathway in the compiler API to resolve. It should have broader coverage, and I believe it should work in your case.

However, I keep getting `Cannot find module '@effect-ts/core/Tracing' from ...`. This seems to apply with v2.2.4 as well. I ran a full build, so I'm not sure why it's happening.

If you get a chance, please let me know if https://github.com/LeDDGroup/typescript-transform-paths/pull/120 works for you

**Comment by mikearnaldi** [comment]
I'll test #120 this evening, what I can say is that I have been using the v2 for a long time and it always worked also within jest. Like the rest of compiler plugins (never seen one where jest generates the issue)

**Comment by nonara** [comment]
I appreciate that. I'll publish after I hear back.

**Comment by mikearnaldi** [comment]
Looks like #120 works partially, it does resolve modules that are present in the original source file but not modules added by other plugins that run before.

Using the following branch: https://github.com/Effect-TS/core/tree/repro-for-ts-path-transform 
In the project root: `yarn && yarn ttsc -p tsconfig.debug.json`

This will build the file `scripts/jest-setup-after-env.ts` that has the following content: 

```ts
import { equals } from "@effect-ts/system/Structural"
import * as U from "jest-matcher-utils"
...
```

When the file is processed, the first plugin of the stack adds an import `import ... from "@effect-ts/core/Tracing"`, this additional module is not resolved.

(Note that I am just using a jest file as example, this repro does not run in jest)

**Comment by nonara** [comment]
Ok. Got it working. A couple points...

**1. Thanks for the excellent repro!** 

Made it very easy to track out. My compliments also on your project structure. Nice work!

**2. You're definitely doing some wild stuff with transformers!**

I can certainly respect that, though, as I'm doing the same in a few of my projects (generated source files, etc)

In this case, the culprit was inserting an import to a file which isn't known to the `Program` instance of the compiler. 

In other words, by default tsc will consume the module for any `import` or `require`. It parses each and creates a `SourceFile`. These are all accessible to the compiler in what's called a `Program` instance. Because you added the import during runtime, and it was to a file which wasn't otherwise referenced, it wasn't known to the compiler API. This is the root to a lot of transformer issues / complaints.

As of v3+, the functionality we're relying on within the compiler requires a `SourceFile`, for a few reasons I won't get into.

What I've done is add functionality to create a skeleton `SourceFile`, in the event that there isn't one within the Program. For speed, it skips the AST parsing, as we only need the outline. This covers your use case.

**3. Regarding your `like the rest of compiler plugins` comment...**

The difference, here, is that most compiler plugins do not rely on the compiler API beyond simple factory functions. Because we want to get the exact resolution strategy used by the compiler, we have to hook into not only undocumented, but in some cases, explicitly *internal* functions of the compiler. As a result, this can be challenging, particularly as it relates to supporting edge cases which go beyond the scope of the general capability of the compiler's functionality.

That said, I believe this issue has helped us to get to a very stable and broadly supported pathway for resolution, so I'm grateful it came up!

Please let me know if this one works for you, and best of luck with your project.

**Comment by mikearnaldi** [comment]
1. Happy to know it was helpful, I also maintain oss libs and I know very well the importance of good repros for issues. Also your project has been of immense help to me and I do use it in prod in many places :)

2. You can say that out loud, it is by all means not a standard setup and up to v3 this was the only project that didn't crash with AST nodes created from previous plugins.

3. Yep I know what you mean I touch the same APIs for my plugins and it is always edgy territory when the AST comes from transformer steps (like the import in this case).

I will test your latest change and report, very happy you decided to tackle this!

**Comment by nonara** [comment]
Fixed in v3.0.2
